jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java

```diff
@@ -44,45 +46,70 @@
   @VisibleForTesting
   static class ClassFileLoader extends ClassLoader {
 
+    /** Maps from class file to class. */
     private final Map<Path, Class<?>> definedClasses = new HashMap<>();
+
     private final Path rootDirectory;
 
     @VisibleForTesting
     ClassFileLoader(Path rootDirectory) {
       this.rootDirectory = rootDirectory;
     }
 
+    /**
+     * Use {@link #findClass(Path)}.
+     *
+     * <p>This method resolves possible dependency classes for the classes loaded from files.
+     */
     @Nullable
     @Override
-    public Class<?> findClass(String name) {
-      return findClass(name, getPathFromClassName(name));
+    public Class<?> findClass(String className) {
+      Path classFile = getPathFromClassName(className);
+
+      if (!Files.exists(classFile)) {
+        // TODO: Log search class failure?
+        // Cannot find corresponding class file. The class is probably in a dependency JAR.
+        // Makes an empty class to prevent NoClassDefFoundError.
+        try {
```

Comment:
This seems very hacky. 

Just to learn a little bit about the problem, the issue here is that the main class isn't in the compiled .class output files and in a dep?

When I was seeing this issue with my dropwizard project I don't think that should have been the case.

Reply:
The old issue was that the classloader couldn't load a class if the things it imports weren't resolvable. For instance some jar dep wasn't loaded when we were searching through the classes. Which is what was hitting you probably with dropwizard.

But we should probably look into this a little more.

Reply:
In this custom classloader, there's two paths for finding a class. The first is resolving a class that is provided in a file, and second is when the JDK is resolving classes that were used in a class loaded from a file. In the second case, classes that don't exist would cause a `NoClassDefFoundError`. However, since all we care is finding the main class in the provided class files, this error doesn't matter, and this change is a way to tell the JDK that it doesn't matter if classes don't exist since all we are looking for is psvm.